### PR TITLE
[ALLUXIO-1694] Add GetFilesPersistInProgressCommand

### PR DESCRIPTION
**Pull request by siyangy**
Add command to list the files that are in progress of async persistence.

https://alluxio.atlassian.net/browse/ALLUXIO-1694

### CODE REVIEW

**Comment by siyangy** [comment]
alluxio-bot, check this please

**Comment by siyangy** [comment]
alluxio-bot, check this pleas

**Comment by siyangy** [comment]
alluxio-bot, check this please

File: core/client/src/main/java/alluxio/client/file/FileSystem.java [hunk 1]
```diff
@@ -390,4 +390,15 @@ void setAttribute(AlluxioURI path, SetAttributeOptions options)
...
   void unmount(AlluxioURI path, UnmountOptions options) throws IOException, AlluxioException;
+
+  /**
+   * Gets all the files with async persistence in progress.
+   *
+   * @return a list of file paths
+   * @throws AlluxioException if an Alluxio error occurs
+   * @throws IOException an I/O error occurs
+   */
+  public List<String> getFilesPersistInProgress()
```
Comment (uronce-cc) [inline core/client/src/main/java/alluxio/client/file/FileSystem.java]: no need for `public`

File: core/client/src/main/java/alluxio/client/file/FileSystem.java [hunk 2]
```diff
@@ -390,4 +390,15 @@ void setAttribute(AlluxioURI path, SetAttributeOptions options)
...
+      throws AlluxioException, IOException;
```
Comment (uronce-cc) [inline core/client/src/main/java/alluxio/client/file/FileSystem.java]: can be put in the same line with the method definition.

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 3]
```diff
@@ -2538,6 +2587,30 @@ private void setAttributeFromEntry(SetAttributeEntry entry)
   }
 
   /**
+   * @return a list of files that are being persisted in the worker
+   */
+  public List<String> getFilesPersistInProgress() {
+    synchronized (mWorkerToAsyncPersistFilesInProgress) {
+      // Removes the files from lost worker.
+      for (WorkerInfo lostWorker : mBlockMaster.getLostWorkersInfo()) {
```
Comment (uronce-cc) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: This is preferred to be put in `LostFilesDetectionHeartbeatExecutor`.

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 4]
```diff
@@ -2612,4 +2685,5 @@ public void close() {
       // Nothing to clean up
     }
   }
+
```
Comment (uronce-cc) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: NIT: remove empty line

File: core/server/src/main/java/alluxio/master/file/FileSystemMasterClientRestServiceHandler.java [hunk 5]
```diff
@@ -79,6 +79,7 @@
   public static final String SCHEDULE_ASYNC_PERSIST = "schedule_async_persist";
   public static final String SET_ATTRIBUTE = "set_attribute";
   public static final String UNMOUNT = "unmount";
+  public static final String GET_FILES_PERSIST_IN_PROGRESS = "files_persist_in_progress";
```
Comment (uronce-cc) [inline core/server/src/main/java/alluxio/master/file/FileSystemMasterClientRestServiceHandler.java]: sort alphabetically

File: core/server/src/main/java/alluxio/master/file/FileSystemMaster.java [hunk 6]
```diff
@@ -226,6 +227,13 @@
   private Future<?> mLostFilesDetectionService;
 
   /**
+   * Map from worker to the files that have been sent to worker for async persistence and are
+   * in progress of async persistence.
+   * */
+  @GuardedBy("itself")
+  private final Map<Long, Set<Long>> mWorkerToAsyncPersistFilesInProgress;
```
Comment (uronce-cc) [inline core/server/src/main/java/alluxio/master/file/FileSystemMaster.java]: Use concurrent map

File: core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java [hunk 7]
```diff
@@ -1060,6 +1061,25 @@ public void workerHeartbeatTest() throws Exception {
...
+   */
+  @Test
+  public void getFilesPersistInProgressTest() throws Exception {
+    createFileWithSingleBlock(ROOT_FILE_URI);
+
+    long fileId = mFileSystemMaster.getFileId(ROOT_FILE_URI);
+    mFileSystemMaster.scheduleAsyncPersistence(ROOT_FILE_URI);
+    Assert.assertTrue(mFileSystemMaster.getFilesPersistInProgress().isEmpty());
+
+    mFileSystemMaster.workerHeartbeat(mWorkerId1, Lists.<Long>newArrayList());
```
Comment (uronce-cc) [inline core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: Use `HeartbeatScheduler` to schedule heartbeat, your current approach may cause flaky test results due to uncertainty of the start and end of heartbeat. See `lostFilesDetectionTest` for an example usage.

File: shell/src/main/java/alluxio/shell/command/GetFilesPersistInProgressCommand.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    List<String> filePaths = mFileSystem.getFilesPersistInProgress();
+    System.out.println("Files that are currently persisted in progress on worker: ");
+    for (String path : filePaths) {
+      System.out.println(path);
+    }
+  }
+
+  @Override
+  public String getUsage() {
+    return "getFilesPersistInProgressCommand";
```
Comment (uronce-cc) [inline shell/src/main/java/alluxio/shell/command/GetFilesPersistInProgressCommand.java]: `getFilesPersistInProgress`

File: shell/src/main/java/alluxio/shell/command/GetFilesPersistInProgressCommand.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+  @Override
+  protected int getNumOfArgs() {
+    return 0;
+  }
+
+  @Override
+  public void run(CommandLine cl) throws IOException, AlluxioException {
+    List<String> filePaths = mFileSystem.getFilesPersistInProgress();
+    System.out.println("Files that are currently persisted in progress on worker: ");
```
Comment (uronce-cc) [inline shell/src/main/java/alluxio/shell/command/GetFilesPersistInProgressCommand.java]: `... that are being persisted.`

File: shell/src/main/java/alluxio/shell/command/GetFilesPersistInProgressCommand.java [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  @Override
+  public String getUsage() {
+    return "getFilesPersistInProgressCommand";
+  }
+
+  @Override
+  public String getDescription() {
+    return "Gets the path of the files that are being async persisted on worker.";
```
Comment (uronce-cc) [inline shell/src/main/java/alluxio/shell/command/GetFilesPersistInProgressCommand.java]: remove `async`

File: tests/src/test/java/alluxio/master/file/FileSystemMasterClientRestApiTest.java [hunk 11]
```diff
@@ -275,4 +276,21 @@ public void unmountTest() throws Exception {
 
     Mockito.verify(mFileSystemMaster).unmount(Mockito.<AlluxioURI>any());
   }
+
+  @Test
+  public void getFilesPersistInProgress() throws Exception {
```
Comment (uronce-cc) [inline tests/src/test/java/alluxio/master/file/FileSystemMasterClientRestApiTest.java]: convention is to end the name with `Test`

File: core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java [hunk 12]
```diff
@@ -1060,6 +1061,25 @@ public void workerHeartbeatTest() throws Exception {
...
+    String filePath = Iterables.getOnlyElement(mFileSystemMaster.getFilesPersistInProgress());
```
Comment (uronce-cc) [inline core/server/src/test/java/alluxio/master/file/FileSystemMasterTest.java]: Is `Iterables` necessary?

**Comment by uronce-cc** [comment]
@siyangy Thanks for the PR! I've done a pass and left some comments.

**Comment by siyangy** [comment]
Per @yupeng9, we are going to close this PR for now.

**Pull request closed without merging**